### PR TITLE
Integrate RouteLineComponent into androidauto

### PR DESCRIPTION
**Pull request by kmadsen**
### Description

Building on top of https://github.com/mapbox/mapbox-navigation-android/pull/5909

https://github.com/mapbox/mapbox-navigation-android/pull/5909 is good to merge now. I worked on resolving all the issues with integrating with `RouteLineComponent` into android auto so there are a few opinions here.

Opening as a draft to share the details. I think the code all works here, but it will need to be broken apart in order to merge into the `androidauto-v0.x.0` release channel. But for testing purposes, this is using the project dependencies directly

This will be removed before merge proposals
```
    api(project(':libnavigation-android'))
    implementation project(":libnavui-app")
    implementation project(":libnavui-maps")
```

### Code sample

End of the day, the CarRouteLine is much simpler when it uses a shared RouteLineComponent

``` kotlin
@MapboxExperimental
@ExperimentalPreviewMapboxNavigationAPI
class CarRouteLine(
    val mainCarContext: MainCarContext,
    private val options: MapboxRouteLineOptions? = null,
    private val contract: RouteLineComponentContract? = null
) : MapboxCarMapObserver {

    private var routeLineComponent: RouteLineComponent? = null

    override fun onAttached(mapboxCarMapSurface: MapboxCarMapSurface) {
        routeLineComponent = RouteLineComponent(
            mapboxMap = mapboxCarMapSurface.mapSurface.getMapboxMap(),
            mapPlugins = mapboxCarMapSurface.mapSurface,
            options = options ?: MapboxRouteLineOptions.Builder(mapboxCarMapSurface.carContext)
                .withRouteLineResources(RouteLineResources.Builder().build())
                .build(),
            contract = contract
        ).also {
            MapboxNavigationApp.registerObserver(it)
        }
    }

    override fun onDetached(mapboxCarMapSurface: MapboxCarMapSurface) {
        routeLineComponent?.let { MapboxNavigationApp.unregisterObserver(it) }
        routeLineComponent = null
    }
}
```

### CODE REVIEW

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/ui/RouteLineComponent.kt [hunk 1]
```diff
@@ -10,61 +10,37 @@ import com.mapbox.maps.plugin.gestures.gestures
...
-import kotlinx.coroutines.flow.stateIn
 import kotlinx.coroutines.launch
 
-@ExperimentalPreviewMapboxNavigationAPI
-interface RouteLineComponentContract {
-    fun setRoutes(mapboxNavigation: MapboxNavigation, routes: List<NavigationRoute>)
-}
-
-@ExperimentalPreviewMapboxNavigationAPI
-internal class MapboxRouteLineComponentContract : RouteLineComponentContract {
```
Comment (abhishek1508) [inline libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/ui/RouteLineComponent.kt:38]: Why did we remove this implementation and it's use from everywhere?

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/ui/RouteLineComponent.kt [hunk 2]
```diff
@@ -10,61 +10,37 @@ import com.mapbox.maps.plugin.gestures.gestures
...
-
 @ExperimentalPreviewMapboxNavigationAPI
 class RouteLineComponent(
     private val mapboxMap: MapboxMap,
     private val mapPlugins: MapPluginProviderDelegate,
     private val options: MapboxRouteLineOptions,
     private val routeLineApi: MapboxRouteLineApi = MapboxRouteLineApi(options),
     private val routeLineView: MapboxRouteLineView = MapboxRouteLineView(options),
-    contractProvider: Provider<RouteLineComponentContract>? = null
+    contract: RouteLineComponentContract? = null
```
Comment (tomaszrybakiewicz) [inline libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/ui/RouteLineComponent.kt:36]: By injecting a `Provider` we defer the creation of a `Contract` until the component is attached. Why the change?

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/ui/RouteLineComponent.kt [hunk 3]
```diff
@@ -75,22 +51,23 @@ class RouteLineComponent(
...
 
-        coroutineScope.launch {
-            mapboxNavigation.flowRouteProgress().collect { routeProgress ->
-                ifNonNull(mapboxMap.getStyle()) { style ->
-                    routeLineApi.updateWithRouteProgress(routeProgress) { result ->
-                        routeLineView.renderRouteLineUpdate(style, result).also {
-                            result.error?.let {
-                                Log.e(TAG, it.errorMessage, it.throwable)
+        when (contract) {
+            is MapboxNavigationObserver -> {
```
Comment (tomaszrybakiewicz) [inline libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/ui/RouteLineComponent.kt:60]: I thought the whole idea behind `Contracts` was to make them simple and to serve as a "glue" code that allows the use of the components in any context.

I don't think components should be responsible for checking if contracts are `MapboxNavigationObservers` and attaching them to `MapboxNavigation` instance. That responsibility should fall on whoever is providing the contract.

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/ActiveRouteLineComponentContract.kt [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    }
+
+    override fun onAttached(mapboxNavigation: MapboxNavigation) {
+        mapboxNavigation.registerRoutesObserver(observer)
+    }
+
+    override fun onDetached(mapboxNavigation: MapboxNavigation) {
+        mapboxNavigation.unregisterRoutesObserver(observer)
+    }
+}
```
Comment (tomaszrybakiewicz) [inline libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/ActiveRouteLineComponentContract.kt:29]: Why are we changing `RouteLineComponentContract` and its default implementation?

This implementation doesn't match previous behavior and will break component installed via "ComponentInstaller.routeLine`.

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/ui/RouteLineComponent.kt (hunk 2 above)
Comment (kmadsen) [reply libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/ui/RouteLineComponent.kt:36]: The contract does not work as it is, so I just made it work and this was not necessary.

Why is the `Provider<` required and a why have lazy initialization required here? We could add `Provider< MapboxRouteLineOptions` as well - but why would we?

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/ui/RouteLineComponent.kt (hunk 1 above)
Comment (kmadsen) [reply libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/ui/RouteLineComponent.kt:37]: This wasn't removed, it was moved to a separate file. This contract of, observing and updating `MapboxNavigation` is also needed by androidauto components. 

So I've called it ActiveRouteLineComponentContract
to have a distinction from PreviewRouteLineComponentContract

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/ui/RouteLineComponent.kt (hunk 2 above)
Comment (kmadsen) [reply libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/ui/RouteLineComponent.kt:36]: It actually, makes more sense for the options to be a provider than the contract. The contract, typically does not have any data associated with it. So why would it have a lazy initialization? The contract itself can define how it provides a definition 🤔

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/ui/RouteLineComponent.kt (hunk 3 above)
Comment (kmadsen) [reply libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/ui/RouteLineComponent.kt:60]: Yeah I'm not in love with this either. Playing with trade-offs.

The current contract has no way to observe the route, it only allows you to set it. Using the RouteLineComponent becomes more complicated downstream if it cannot observe the routes in the contract

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/ActiveRouteLineComponentContract.kt (hunk 4 above)
Comment (kmadsen) [reply libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/ActiveRouteLineComponentContract.kt:29]: yeah I'm trying some other ways. But the current contract does not allow for observing routes, only setting them.

I was changing the RouteLineComponentContract to make it simpler down stream 🤷 

I can leave it, and handle the oddities in android auto and see what it looks like

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/ui/RouteLineComponent.kt (hunk 3 above)
Comment (abhishek1508) [reply libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/ui/RouteLineComponent.kt:60]: Isn't this where the other ticket I mentioned comes into play. When setting the routes using the contract, the customer might be interested in knowing about this change. https://github.com/mapbox/navigation-sdks/issues/1436 and https://github.com/mapbox/navigation-sdks/issues/1645. 
I think this is what you are looking for.

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/ActiveRouteLineComponentContract.kt (hunk 4 above)
Comment (kmadsen) [reply libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/ActiveRouteLineComponentContract.kt:29]: Yeah, it is a bit more difficult to use `RouteLineComponent` without changing it. But here is the issue I'm not sure how to handle yet

The MapboxNavigation alternative routes are selected by re-ordering the list, in order to select routes without re-ordering the list requires a duplicate route provider for the camera and route line

When selecting the Franklin Street. The camera bounds needs to update to that route, and the list should not be re-ordered.

Screen Shot 2022-06-08 at 3 17 33 PM

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/ui/RouteLineComponent.kt (hunk 3 above)
Comment (kmadsen) [reply libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/ui/RouteLineComponent.kt:60]: I think the issue is different.. But there are a few approaches

**Comment by kmadsen** [comment]
Closing this one in favor of this one https://github.com/mapbox/mapbox-navigation-android/pull/5912
